Base/Python/slicer/parameterNodeWrapper/serializers.py

```diff
@@ -819,3 +821,147 @@ def supportsCaching(self):
+        return None
+
+    def __init__(self, serializers):
+        # This class depends on ValidatedSerializer methods, so we wrap all the serializers in them.
+        # ValidatedSerializers collapse into each other, so it is fine if the serializer is already
+        # a validated serializer.
+        self._serializers = [ValidatedSerializer(s, []) for s in serializers]
+
+    def default(self):
+        if len(self._serializers) == 2 and type(self._serializers[1].serializer) == NoneSerializer:
+            # Special case for if someone does typing.Optional[X] we default to None.
+            # In Python, there is no difference between typing.Optional[X] and typing.Union[X, None],
+            # which is somewhat unfortunate for semantic reasons.
+            return None
+        else:
+            return self._serializers[0].default()
+
+    @staticmethod
+    def _paramName(name, index):
+        return f"{name}_{index}"
+
+    def isIn(self, parameterNode, name: str) -> bool:
+        for index, validatedSerializer in enumerate(self._serializers):
+            if validatedSerializer.isIn(parameterNode, self._paramName(name, index)):
+                return True
+        return False
+
+    def write(self, parameterNode, name: str, value) -> None:
+        with slicer.util.NodeModify(parameterNode):
+            wasIn = self.isIn(parameterNode, name)
+            oldValue = self.read(parameterNode, name) if wasIn else None
+            raisedException = None
+            try:
+                self.remove(parameterNode, name)
+                for index, validatedSerializer in enumerate(self._serializers):
+                    # canSerialize does a good job most of the time at differentiating between different
+                    # types of the union.
+                    # The exception is if there are multiple MRML nodes of different types.
+                    # The NodeSerializer staticmethod canSerialize doesn't know which particular node type
+                    # the instance is ready for. So let's try and if we fail, see if there is another that
+                    # succeeds.
+                    if validatedSerializer.serializer.canSerialize(type(value)):
+                        try:
+                            validatedSerializer.validate(value)
+                            validatedSerializer.write(parameterNode, self._paramName(name, index), value)
+                            break
+                        except Exception as e:
+                            raisedException = e
+                else:
+                    if raisedException is not None:
+                        raise raisedException
+                    raise TypeError(f"Unable to write value '{value}' of type '{type(value)}'")
+            except Exception:
+                if wasIn:
+                    self.write(parameterNode, name, oldValue)
+                else:
+                    self.remove(parameterNode, name)
+                raise
```

Comment:
Really not a fan of this logic... Nested `try`-`except Exception` seems problematic.

I'm not certain that `wasIn`, `oldValue`, `raisedException` are strictly necessary. `validatedSerializer.write` will not be called unless `validatedSerializer.validate()` succeeds; and if an error occurs in `validatedSerializer.write` then it seems appropriate to let that pass through.

If I understand the API correctly, this is equivalent:

```python
def write(self, parameterNode, name: str, value) -> None:
    with slicer.util.NodeModify(parameterNode):
        for index, serializer in enumerate(self._serializers):
            if not serializer.canSerialize(type(value)):
                continue

            try:
                serializer.validate(value)
            except Exception:
                continue

            self.remove(parameterNode, name)
            serializer.write(parameterNode, self._paramName(name, index), value)

            return

    raise TypeError(f"Unable to write value {value!r} of type {type(value)}")
```

It might be appropriate to _require_ `validate()` to raise `ValueError` for failed validation as part of the API, else the exception is regarded as a true error and is not caught.

Could also let `validate()` return `bool` and use it as a predicate.

I'm not sure of the best pattern to choose, since exceptions allow to attach some error message; either way, it'd still be nice to tighten or remove those `except` clauses.

Reply:
The code snippet I included is _not_ equivalent; I'll look at it some more and update if I come up with a better approach.

Reply:
The code above fails tests because they expect `TypeError` to be raised if all `canSerialize` fail; but they expect `ValueError` to be raised if `validate` fails.

It's unclear what to expect if _multiple_ `validate()` fail. We do not have `ExceptionGroup` in Python 3.9.

Reply:
Discussed offline with Connor; there are enough edge cases that the fundamental architecture will be unchanged, but add a separate method for encapsulation.